pkg/operator/controller/controller_router_deployment.go

```diff
@@ -0,0 +1,226 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+
+	"github.com/sirupsen/logrus"
+
+	ingressv1alpha1 "github.com/openshift/cluster-ingress-operator/pkg/apis/ingress/v1alpha1"
+	"github.com/openshift/cluster-ingress-operator/pkg/manifests"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	configv1 "github.com/openshift/api/config/v1"
+)
+
+// routerDeploymentName returns the namespaced name for the router deployment.
+func routerDeploymentName(ci *ingressv1alpha1.ClusterIngress) types.NamespacedName {
+	return types.NamespacedName{
+		Namespace: "openshift-ingress",
+		Name:      "router-" + ci.Name,
+	}
+}
+
+// ensureRouterDeployment ensures the router deployment exists for a given
+// clusteringress.
+func (r *reconciler) ensureRouterDeployment(ci *ingressv1alpha1.ClusterIngress, infraConfig *configv1.Infrastructure, ingressDomain *string, ha *ingressv1alpha1.ClusterIngressHighAvailability) (*appsv1.Deployment, error) {
+	if infraConfig == nil || ha == nil {
+		return nil, nil
+	}
+
+	expected, err := desiredRouterDeployment(ci, r.Config.RouterImage, ingressDomain, infraConfig, ha.Type)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build router deployment: %v", err)
+	}
+	current := expected.DeepCopy()
+	err = r.Client.Get(context.TODO(), types.NamespacedName{Namespace: expected.Namespace, Name: expected.Name}, current)
```

Comment:
Might as well introduce `currentRouterDeployment` while you're at it?

Reply:
Done.